projects/hslayers/src/common/layers/interpolated-source.ts

```diff
@@ -0,0 +1,200 @@
+                const toRemove = extentCache.filter(
+                  //Delete cached extents which contain this extent because they have their feature counts limited
+                  (cachedExt) =>
+                    !equals(cachedExt, extent) &&
+                    containsExtent(cachedExt, extent)
+                );
+                for (const extToRemove of toRemove) {
+                  super.getSource().removeLoadedExtent(extToRemove);
+                }
+                return [extent];
+              },
+        loader: async (extent, resolution, projection, success, failure) => {
+          if (options.loader) {
+            this.fillFeatures(
+              await options.loader({
+                extent,
+                resolution,
+                projection,
+                success,
+                failure,
+              }),
+              extent
+            );
+          }
+        },
+      }),
+      weight: NORMALIZED_WEIGHT_PROPERTY_NAME,
+    });
+    if (options.colorMap) {
+      if (typeof options.colorMap == 'string') {
+        super.getColor = (v) => {
+          const black = [0, 0, 0, 255];
+          if (isNaN(v)) {
+            return black;
+          }
+          if (v > 99) {
+            v = 99;
+          }
+          if (v < 0) {
+            v = 0;
+          }
+          v = Math.floor(v);
+          return colorMaps[options.colorMap as string][v];
+        };
+      } else {
+        super.getColor = options.colorMap;
+      }
+    }
+    if (options.features) {
+      this.fillFeatures(options.features);
+    }
+  }
+
+  /**
+   * Fill Interpolated source features
+   * @param features - Parsed Ol features from get request
+   * @param extent - Current map extent
+   */
+  fillFeatures(features: Feature<Geometry>[], extent?: number[]) {
+    if (!features) {
+      return;
+    }
+    const featsCached = this.featureCache.getFeatures();
+    const cacheLimit = this.options.maxFeaturesInCache;
+    if (cacheLimit < featsCached.length + features.length) {
+      const cntToRemove = featsCached.length - (cacheLimit - features.length);
+      featsCached
+        .slice(0, cntToRemove)
+        .forEach((f) => this.featureCache.removeFeature(f));
+    }
+    const countToAdd = cacheLimit ?? Number.MAX_VALUE;
+    this.featureCache.addFeatures(features.slice(0, countToAdd));
+    this.normalizeWeight(this.options.weight);
+    const src = super.getSource();
+    if (extent) {
+      src.clear();
+      const limitInExt = this.options.maxFeaturesInExtent ?? Number.MAX_VALUE;
+      src.addFeatures(
+        this.featureCache.getFeaturesInExtent(extent).slice(0, limitInExt)
+      );
+    } else {
+      src.addFeatures(features);
+    }
+  }
+
+  /**
+   * Parse features from get request
+   * @param collection - Get request response feature collection
+   * @param mapProjection - Map projection
+   */
+  parseFeatures(collection: any, mapProjection: string | Projection): void {
```

Comment:
`collection: any` but then we rely on its `features` property. Not much safety in that...

Reply:
Can You suggest any better type for it?

Reply:
From the usage inside the method, it would be obviously
```
{
  crs?: string;
  srs?: string;
  features: string[];
}
```

but given the actual use of this method, which we do not use internally but is only called from the loader in the test-app, I would say moving the if-condition from there to here might be sufficient (and keeping the 'any').
I.e. the line `if (response?.features?.length > 0) {` gives me much more sense inside here than in every custom loader a consumer of hslayers would need to write.

Does it make sense to you? 😄